### PR TITLE
make O_DIRECT optional

### PR DESCRIPTION
**Pull request by nraynaud**
- retry open() with O_DSYNC if an EINVAL error was encountered

ZFS doesn't support O_DIRECT, so I adopted what I hope is the least intrusive changes to the codebase.

### CODE REVIEW

**Comment by nraynaud** [comment]
Hello, 
can you explain your process and what it would take to merge this change please?
I just signed off my commit.

Thanks, 

Nicolas.

**Comment by BobBall** [comment]
It'd be cleaner to refactor the open calls to a new method which will try first with the flags passed, then remove O_DIRECT and replace with O_DSYNC in the failure case.  Can we also get some logging when we switch from the default flags so we can see what's going on?

**Comment by nraynaud** [comment]
@BobBall sure, can you suggest a name and a residing file for the new function?

**Comment by nraynaud** [comment]
@BobBall I did my best to comply. 

What's next in the quest to get this merged?

File: vhd/lib/libvhd.c [hunk 1]
```diff
@@ -117,6 +118,24 @@ old_clear_bit(volatile char *addr, int nr)
 	((uint32_t *)addr)[nr >> 5] &= ~(1 << (nr & 31));
 }
 
+int
+open_optional_odirect(const char *pathname, int flags, ...)
+{
+	mode_t mode;
+	if (flags & O_CREAT) {
```
Comment (BobBall) [inline vhd/lib/libvhd.c]: I wonder if this complexity is worth it - it'd be easier to make the wrapper always take a mode_t (which could be passed as -1 in the calls which are not O_CREAT) - although if Mark is happy then I have no objections.

**Comment by nraynaud** [comment]
I have passed 0, it feels like the safest value.

**Comment by BobBall** [comment]
Thanks - I'm happy.

**Pull request merged**